Update SiteApplication.php to initialize properly the template object caused by #30192

joeforjoomla opened the pull request:
Related to #30192 

### Summary of Changes
Add the new properties 'parent' and 'inheritable' to the object initialized in the 'setTemplate' method so that there are no 'Notice' if a plugin calls the 'setTemplate' method to override the template used


### Testing Instructions
Call the 'setTemplate' method from a system plugin


### Actual result BEFORE applying this Pull Request
Notice for undefined property 'parent'
![image](https://user-images.githubusercontent.com/17835460/90122019-0099d500-dd5d-11ea-9393-0cbcf8e8689e.png)



### Expected result AFTER applying this Pull Request
No PHP notice


### Documentation Changes Required
No

joeforjoomla committed:
Update SiteApplication.php

joeforjoomla commented:
@dgrammatiko you are here

dgrammatiko commented:
I'll do a PR unless you are faster

dgrammatiko commented:
@joeforjoomla sure something I overlooked, but I guess hardcoding the params to null and 0 will only work for the legacy mode. The function should be like:
```php
	public function setTemplate($template, $styleParams = null, $inheritable = 0, $parent = '')
	{...}
```
So it can work for all cases

joeforjoomla commented:
Yes @dgrammatiko actually the setTemplate should have new parameters t oallow setting both new properties. That would be ideal.

joeforjoomla committed:
Update SiteApplication.php

joeforjoomla commented:
@joeforjoomla i've updated this PR accordingly

wilsonge committed:
Codestyle

Quy commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -864,18 +864,22 @@ public function setLanguageFilter($state = false)
 	 * @since   3.2
 	 */
-	public function setTemplate($template, $styleParams = null)
+	public function setTemplate($template, $styleParams = null, inheritable = 0, $parent= null)
```
```suggestion
	public function setTemplate($template, $styleParams = null, $inheritable = 0, $parent= null)
```

Quy commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -864,18 +864,22 @@ public function setLanguageFilter($state = false)
 			$this->template = new \stdClass;
 			$this->template->template = $template;
-
+			$this->template->inheritable = inheritable;
```
```suggestion
			$this->template->inheritable = $inheritable;
```

dgrammatiko committed:
I was wrong

Both inheritable and parent are params from the template_styles table, so it makes sense to keep them consistent

dgrammatiko commented:
@joeforjoomla @wilsonge I was wrong here, we don't need to change the signature of the function, check https://github.com/joeforjoomla/joomla-cms/pull/1

wilsonge committed:
Update libraries/src/Application/SiteApplication.php

Co-authored-by: Quy <quy@fluxbb.org>

wilsonge committed:
Update libraries/src/Application/SiteApplication.php

Co-authored-by: Quy <quy@fluxbb.org>

richard67 commented:
> ### Testing Instructions
> 
> Call the 'setTemplate' method from a system template

@joeforjoomla From a system template? Or from a system plugin?

joeforjoomla commented:
@richard67 ops sorry.. obviously system plugin

richard67 commented:
No problem, can happen.

richard67 commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -864,18 +864,22 @@ public function setLanguageFilter($state = false)
-
+			$this->template->inheritable = $inheritable;
+			$this->template->parent = $parent;
+			
```
Please remove white space in this line, but leave this empty line, to fix PHP code style errors as reported by Drone (one of our CI tools) here: [https://ci.joomla.org/joomla/joomla-cms/34685/1/9](https://ci.joomla.org/joomla/joomla-cms/34685/1/9).

Thanks in advance.

richard67 replied:
I would have made a suggestion for a one click, but I've made the experience that GitHub then removes the complete line, and that would also not be according to our CS.

wilsonge committed:
Whitespace

richard67 commented:
Unit test failure is not related to this PR.

joeforjoomla committed:
Merge branch 'patch-2' into patch-1

joeforjoomla committed:
Merge pull request #1 from dgrammatiko/patch-1

I was wrong

SharkyKZ commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -875,10 +875,22 @@ public function setTemplate($template, $styleParams = null)
 			{
 				$this->template->params = $styleParams;
+				
+				if (isset($styleParams->inheritable))
```
This will never be defined.

dgrammatiko replied:
Why?

SharkyKZ replied:
Shouldn't stuff random properties on an object like that. `$styleParams` is supposed to be `Joomla\Registry\Registry` object with template params loaded in.

SharkyKZ replied:
You could probably get away with changing the method to accept the entire "template" object like described https://github.com/joomla/joomla-cms/issues/30314 here.

JoeJoomlaCoder commented:
I have tested this item :white_check_mark: successfully on 578166f2d434fd61212fcd370b0284d7092ec78f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30360">issues.joomla.org/tracker/joomla-cms/30360</a>.</sub>

joeforjoomla commented:
RTC

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30360">issues.joomla.org/tracker/joomla-cms/30360</a>.</sub>

infograf768 commented:
Back to pending as there are unaddressed comments.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30360">issues.joomla.org/tracker/joomla-cms/30360</a>.</sub>

richard67 commented:
@joeforjoomla Is @JoeJoomlaCoder also an account from you?

rmittl commented:
I have tested this item :white_check_mark: successfully on 578166f2d434fd61212fcd370b0284d7092ec78f<br /><br />Called setTemplate() method in a System Plugin -> PHP Notice
After applied Patch -> no PHP Notice Warning<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30360">issues.joomla.org/tracker/joomla-cms/30360</a>.</sub>

rmittl commented:
Called setTemplate() method in a System Plugin -> PHP Notice
After applied Patch -> no PHP Notice Warning
Tested successfully<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30360">issues.joomla.org/tracker/joomla-cms/30360</a>.</sub>

richard67 commented:
This PR still has open review comments and also fail the PHP code style test by drone.

For fixing the latter I'll suggest a correction in a minute.

But as long as the review comments are not addressed I can't really set RTC.

gostn commented:
Where come this double comments from:

![image](https://user-images.githubusercontent.com/71960554/96335199-49c91680-1077-11eb-9bb4-55abfbc25bb8.png)

richard67 commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -875,10 +875,22 @@ public function setTemplate($template, $styleParams = null)
 			if ($styleParams instanceof Registry)
 			{
 				$this->template->params = $styleParams;
+				
```
```suggestion

```
Fix PHP code style.

rmittl commented:
Sorry, my fault. The issues Joomla Site update not correctly after posting "Test this". So I post it again and it was double.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30360">issues.joomla.org/tracker/joomla-cms/30360</a>.</sub>

joeforjoomla committed:
Update libraries/src/Application/SiteApplication.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -875,10 +875,21 @@ public function setTemplate($template, $styleParams = null)
 			if ($styleParams instanceof Registry)
 			{
 				$this->template->params = $styleParams;
+				if (isset($styleParams->inheritable))
```
```suggestion

				if (isset($styleParams->inheritable))
```
Sorry, it seems with the last correction GitHub has removed the complete empty line and not the spaces within that line. But that empty line is needed. Now PHP code style failed again. Just apply again the suggestion and this time it will be fine, I promise ;-)

joeforjoomla committed:
Update libraries/src/Application/SiteApplication.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

joeforjoomla commented:
@richard67 any update on this? Can we get it merged quickly? What is still missing?

dgrammatiko commented:
@joeforjoomla please address this comment https://github.com/joomla/joomla-cms/pull/30360#discussion_r475080728

joeforjoomla commented:
@dgrammatiko i don't think that #30360 is a good idea, having a string or an object assigned to the property $this->template->template is not an option, indeed extension have always assumed that this is a string matching the template name. However #30360 is quite unrelated to this PR.

dgrammatiko commented:
> having a string or an object assigned to the property $this->template->template is not an option

I guess that was an example code and they meant `$this->template = $template`.  The `$this->template->template` is the value of the db from the table extensions and it's a string. Other than that the suggestion is solid and tbh expected behaviour

joeforjoomla commented:
Ok but honestly i don't understand what i should do here in relation to this PR. Eventually that can be merged as a separate PR.

joeforjoomla commented:
This is still an issue. What to do to merge a fix?

brianteeman commented:
Creating a fake account in order to mark that you had tested something doesnt really inspire people to accept that test result.

If you want people to test something that is not trivial then you need to give test instructions that are easy to follow

joeforjoomla commented:
@brianteeman this is an open bug, nobody is interested to create fake accounts but just fix the Joomla code

rdeutz committed:
Update SiteApplication.php

small change to let ci running

richard67 commented:
Closing in favour of #32710 . @joeforjoomla Sorry, I hope you are not too angry or disappointed. Feel free to re-open here if you think it is a mistake that I closed it.
